renderdoc/driver/shaders/dxbc/dxbc_debug.cpp

```diff
@@ -1861,6 +1861,40 @@ void FlattenVariables(const rdcstr &cbufferName, const rdcarray<ShaderConstant>
   }
 }
 
+void ThreadState::MarkResourceAccess(ShaderDebugState *state, DXBCBytecode::OperandType type,
+                                     const BindingSlot &slot)
+{
+  if(state == NULL)
+    return;
+
+  if(type != DXBCBytecode::TYPE_RESOURCE && type != DXBCBytecode::TYPE_UNORDERED_ACCESS_VIEW)
+    return;
+
+  state->changes.push_back(ShaderVariableChange());
+  ShaderVariableChange &change = state->changes.back();
+  change.before.type = (type == DXBCBytecode::TYPE_RESOURCE) ? VarType::ReadOnlyResource
+                                                             : VarType::ReadWriteResource;
+
+  uint32_t reg = slot.shaderRegister;
+  uint32_t arrIdx = 0;
+
+  const rdcarray<DXBC::ShaderInputBind> &shaderBinds =
+      (type == DXBCBytecode::TYPE_RESOURCE) ? reflection->SRVs : reflection->UAVs;
+  for(size_t i = 0; i < shaderBinds.size(); ++i)
+  {
+    const DXBC::ShaderInputBind &bind = shaderBinds[i];
+    if(bind.space == slot.registerSpace && bind.reg <= slot.shaderRegister &&
+       (bind.bindCount == ~0U || slot.shaderRegister <= bind.reg + bind.bindCount))
+    {
+      reg = bind.reg;
+      arrIdx = slot.shaderRegister - bind.reg;
+      break;
+    }
+  }
+  change.before.SetBinding(slot.registerSpace, reg, arrIdx);
+  change.after = change.before;
```

Comment:
For a new register coming into scope, the `before` needs to be a default-initialised `ShaderVariable()`. That way backwards iteration knows to remove it again when stepping back over this step.